lib/dynamoid/adapter_plugin/aws_sdk_v2.rb

```diff
@@ -559,6 +569,15 @@ def scan(table_name, scan_hash, select_opts = {})
           record_count = 0
           scan_count = 0
           loop do
+            # Adjust the limit down if the remaining record or scan limit are
+            # lower to obey limits. We can assume the difference won't be
+            # negative due to break statements below.
+            if request[:limit] && record_limit && record_limit - record_count < request[:limit]
+              request[:limit] = record_limit - record_count
+            elsif request[:limit] && scan_limit && scan_limit - scan_count < request[:limit]
+              request[:limit] = scan_limit - scan_count
+            end
+
```

Comment:
>  SCAN will only scan through maximum of 1MB of data

it's documented behaviour - http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.html#Query.Pagination

> I don't think this is the case with limit: 1

To be clear I meant `record_limit`. Okay. Let's consider an example.

User sets `record_limit` and some conditions, like `User.where('created_at.gt' => Time.now - 1.day).record_limit(1000)`

First api call - `request[:limit]` is 1000
1000 items were scanned and only 999 were filtered and returned

Next api call - `request[:limit]` is 1000-999=1

So we will make several api calls with this small limit (1) until find one matched item